distsql: Check for cancellation while waiting for inbound streams

abhimadan opened the pull request:
Previously, the flow context was checked for cancellation just before
the server sent an RPC to producers. However, if the context was
canceled while blocked on the RPC, the cancellation wouldn't be picked
up until the next producer RPC was sent.

To avoid this problem, the RPC is now sent in its own goroutine, and the
server waits for either context cancellation or the RPC to return,
whichever comes first.

Release note (performance improvement): Actively push cancellations up
DistSQL flows, without waiting for the first row.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22737)
<!-- Reviewable:end -->

asubiotto commented:
Thanks for doing this! @RaduBerinde would you mind taking a quick look?

Could you run some benchmarks so that we can see the performance impact of this change in general? Maybe `Benchmark{Count,Sort}/MultinodeCockroach` in `pkg/bench` would be useful here. Since you already have a sf10 cluster set up, maybe also try running a long-running query and checking out the difference.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/sql/distsqlrun/inbound.go, line 74 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L5PRwTC2CngMORvlNFp:-L5PRwTC2CngMORvlNFq:b-kaa17z) ([raw file](https://github.com/cockroachdb/cockroach/blob/089dbf4d74ec7aa720dc10909d79930869d35ec0/pkg/sql/distsqlrun/inbound.go#L74)):*
> ```Go
> 			errChan := make(chan error, 1)
> 			doneRPC := make(chan struct{}, 1)
> 			go func() {
> ```

I'm not sure we want to create a goroutine on each iteration of this loop. I think this would still work if we create it at the start of this function with the channels.

---

*[pkg/sql/distsqlrun/inbound.go, line 78 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L5PSAo-D6X94V-THJZb:-L5PSAo-D6X94V-THJZc:b-jikhub) ([raw file](https://github.com/cockroachdb/cockroach/blob/089dbf4d74ec7aa720dc10909d79930869d35ec0/pkg/sql/distsqlrun/inbound.go#L78)):*
> ```Go
> 				msgChan <- prodMsg
> 				errChan <- err
> 				doneRPC <- struct{}{}
> ```

We can probably get rid of this channel and use the errChan for the same purpose as `doneRPC` then doing the errCheck in the `case` below.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L5PU4YY0LGWQrnoFy8s:b-2d7azm)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/inbound.go, line 131 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L5PVdN37Ec5smqGNYl9:-L5PVdN37Ec5smqGNYlA:b-2i3fcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/089dbf4d74ec7aa720dc10909d79930869d35ec0/pkg/sql/distsqlrun/inbound.go#L131)):*
> ```Go
> 			case NeedMoreRows:
> 				continue
> 			case DrainRequested:
> ```

Just commenting here because I was confused about this. I was wondering what the difference between returning an error and sending a `DrainRequest` was since I only see a `flowCtxCancel` when handling a drain request on the outbox. It seems that `FlowStream` (to set up this stream) is asynchronous and returning an error here actually sends it over the stream (so it's not a call failure). The outbox then listens for this error in `listenForDrainSignalFromConsumer` and properly initiates a drain sequence.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
On my TPC-H cluster with scalefactor 10 data, the query `SELECT c_name, MAX(c_acctbal) FROM tpch.customer GROUP BY c_name ORDER BY c_name` is canceled almost instantaneously. I unfortunately don't have any measurements from before my change, but in #22650, a similar query took 15s to cancel.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/inbound.go, line 74 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L5PRwTC2CngMORvlNFp:-L5PpjZP2e2kVo2OmClp:begz8n2) ([raw file](https://github.com/cockroachdb/cockroach/blob/089dbf4d74ec7aa720dc10909d79930869d35ec0/pkg/sql/distsqlrun/inbound.go#L74)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm not sure we want to create a goroutine on each iteration of this loop. I think this would still work if we create it at the start of this function with the channels.
</blockquote></details>

If it's created at the start of this function, then the goroutine will need extra synchronization logic with the main loop in this function. I don't feel strongly either way, but I was just wondering if that was what you had in mind.

---

*[pkg/sql/distsqlrun/inbound.go, line 78 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L5PSAo-D6X94V-THJZb:-L5Ppiog15CjKObzTQKX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/089dbf4d74ec7aa720dc10909d79930869d35ec0/pkg/sql/distsqlrun/inbound.go#L78)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

We can probably get rid of this channel and use the errChan for the same purpose as `doneRPC` then doing the errCheck in the `case` below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L5Pp3H4Ap7L64gwrN9I:b15x9ab)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/inbound.go, line 74 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L5PRwTC2CngMORvlNFp:-L5Prxg97_1WClYlSXmP:b9kxslk) ([raw file](https://github.com/cockroachdb/cockroach/blob/089dbf4d74ec7aa720dc10909d79930869d35ec0/pkg/sql/distsqlrun/inbound.go#L74)):*
<details><summary><i>Previously, abhimadan (Abhishek Madan) wrote…</i></summary><blockquote>

If it's created at the start of this function, then the goroutine will need extra synchronization logic with the main loop in this function. I don't feel strongly either way, but I was just wondering if that was what you had in mind.
</blockquote></details>

Never mind, the synchronization logic was fairly straightforward, and I think your point about creating a goroutine once per loop iteration is more important than simplicity anyway. Made the change.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/inbound.go, line 71 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L5Psm7i1kLbukKpVkT-:-L5Psm7i1kLbukKpVkT0:b-2we6gh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d7ba7cf47bf37900d0263730a464f273325994a/pkg/sql/distsqlrun/inbound.go#L71)):*
> ```Go
> 	defer func() { doneChan <- true }()
> 
> 	go func() {
> ```

If we get cancelled, can't this goroutine hang around inside Recv for an arbitrary amount of time? Server shutdown won't wait for it, which could lead to flaky tests (goroutine leak errors).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
So it seems that this is fast only because `dst.ProducerDone()` is called before the inbound stream's producers are marked as complete. This doesn't seem to align with the way completion is usually marked, by starting at the data sources and working towards the `syncFlowConsumer`. As a result, this approach is probably incorrect for more reasons than the goroutine leak. I'm going to think some more on how to approach this in the correct way.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L5R7qKkCeMLgnEodgPp:b-yawxqy)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
So my original comment wasn't entirely correct; the error returned from `ProcessInboundStream` is sent as an RPC error to the FlowStreamClient in the corresponding outbox, which resulted in cancelling the context on that node. However, the inbound stream doesn't return as soon as it detects a cancellation now, so we needed a new approach.

I added a new `CancelRequest` that FlowStream consumers can send to producers, which indicates that the producer should cancel their flow context. The RPC stream is also closed by the outbox to unblock the inbound consumer stream blocked on a `Recv()` call.

I'm fairly confident that this works now; cancelling the query `SELECT l_shipmode, AVG(l_extendedprice) FROM tpch.lineitem GROUP BY l_shipmode` on a 5-node roachprod cluster with TPC-H scalefactor 1 data is almost instantaneous. The goroutine leak is also gone now. PTAL!

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L63Cw0P1tsgka5E3--z:bohtbjg)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/inbound.go, line 131 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L5PVdN37Ec5smqGNYl9:-L63EOi315f55M7keQl8:bi31ca3) ([raw file](https://github.com/cockroachdb/cockroach/blob/089dbf4d74ec7aa720dc10909d79930869d35ec0/pkg/sql/distsqlrun/inbound.go#L131)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Just commenting here because I was confused about this. I was wondering what the difference between returning an error and sending a `DrainRequest` was since I only see a `flowCtxCancel` when handling a drain request on the outbox. It seems that `FlowStream` (to set up this stream) is asynchronous and returning an error here actually sends it over the stream (so it's not a call failure). The outbox then listens for this error in `listenForDrainSignalFromConsumer` and properly initiates a drain sequence.
</blockquote></details>

A non-EOF RPC error tells the outbox to fail fast it won't send any metadata or talk to the inbound stream at all afterwards (since it's no longer listening). A `DrainRequest`, on the other hand, implies that the inbound stream is still waiting for metadata, so the outbox doesn't stop talking to its RowChannel. These cases are managed by the `drainRequest.err != nil` check in `outbox.go`; the error is only set on a non-EOF RPC error (and now when it receives a `CancelRequest`).

With this change, the new `CancelRequest` is sort of a middle ground between non-EOF RPC errors and `DrainRequests`: it doesn't want metadata, but it wants the outbox to close the stream to unblock the inbound stream's RPC call.

---

*[pkg/sql/distsqlrun/inbound.go, line 71 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L5Psm7i1kLbukKpVkT-:-L63EKFQ-hUkV9HotcDK:bkl1ejk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d7ba7cf47bf37900d0263730a464f273325994a/pkg/sql/distsqlrun/inbound.go#L71)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

If we get cancelled, can't this goroutine hang around inside Recv for an arbitrary amount of time? Server shutdown won't wait for it, which could lead to flaky tests (goroutine leak errors).
</blockquote></details>

The leak should be gone now.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Inventing a new signal from consumer to producer and involving the producer to unblock the consumer seems unfortunate to me. Would it work if we did the `stream.Recv()` call on another goroutine than the one servicing the `processInboundStreamHelper()` call? So the RPC goroutine would return as soon as ctx cancelation is detected, and that should lead to the `Recv()` goroutine also unblocking soon after. That would seem to me more like how God intended things: the RPC client says that it's done by returning from the handler, not by inventing its own protocol. I guess the question is whether gRPC minds returning from the RPC handler while someone is Recv()ing on the stream.

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/inbound.go, line 92 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L63H_Qn0k0q21Jq6orA:-L63H_Qn0k0q21Jq6orB:b4w8cq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa515f0e8070d3a792d8e9b6b546df981837a7d/pkg/sql/distsqlrun/inbound.go#L92)):*
> ```Go
> 						err, log.MakeMessage(ctx, "communication error", nil /* args */))
> 				}
> 				// The stream has closed and the flow context is canceled. This
> ```

nit: either say "If the stream...", or move the comment inside the `if`.

---

*[pkg/sql/distsqlrun/inbound.go, line 174 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L63HFy_4sGksyXjVfGQ:-L63HFy_4sGksyXjVfGR:b-uxx39f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa515f0e8070d3a792d8e9b6b546df981837a7d/pkg/sql/distsqlrun/inbound.go#L174)):*
> ```Go
> 	log.VEvent(ctx, 1, "sending cancel signal to producer")
> 	sig := ConsumerSignal{CancelRequest: &CancelRequest{}}
> 	return stream.Send(&sig)
> ```

can this race with `sendDrainSignalToStreamProducer()`? Is that OK?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L63FcP-26NzlRvlR8Ch:bxlz6qh)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Btw, I should have said that I'm very glad you're doing something about
cancelation in the inbound stream processing. I had recently noticed how we
block on the Recv call and I though we should figure out a better story.

On Fri, Feb 23, 2018 at 3:20 PM, Andrei Matei <notifications@github.com>
wrote:

> Inventing a new signal from consumer to producer and involving the
> producer to unblock the consumer seems unfortunate to me. Would it work if
> we did the stream.Recv() call on another goroutine than the one servicing
> the processInboundStreamHelper() call? So the RPC goroutine would return
> as soon as ctx cancelation is detected, and that should lead to the Recv()
> goroutine also unblocking soon after. That would seem to me more like how
> God intended things: the RPC client says that it's done by returning from
> the handler, not by inventing its own protocol. I guess the question is
> whether gRPC minds returning from the RPC handler while someone is
> Recv()ing on the stream.
> ------------------------------
>
> Review status: 0 of 4 files reviewed at latest revision, 4 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *pkg/sql/distsqlrun/inbound.go, line 92 at r3
> <https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L63H_Qn0k0q21Jq6orA:-L63H_Qn0k0q21Jq6orB:b4w8cq0>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/3fa515f0e8070d3a792d8e9b6b546df981837a7d/pkg/sql/distsqlrun/inbound.go#L92>):*
>
> 						err, log.MakeMessage(ctx, "communication error", nil /* args */))
> 				}
> 				// The stream has closed and the flow context is canceled. This
>
> nit: either say "If the stream...", or move the comment inside the if.
> ------------------------------
>
> *pkg/sql/distsqlrun/inbound.go, line 174 at r3
> <https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L63HFy_4sGksyXjVfGQ:-L63HFy_4sGksyXjVfGR:b-uxx39f>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/3fa515f0e8070d3a792d8e9b6b546df981837a7d/pkg/sql/distsqlrun/inbound.go#L174>):*
>
> 	log.VEvent(ctx, 1, "sending cancel signal to producer")
> 	sig := ConsumerSignal{CancelRequest: &CancelRequest{}}
> 	return stream.Send(&sig)
>
> can this race with sendDrainSignalToStreamProducer()? Is that OK?
> ------------------------------
>
> *Comments from Reviewable
> <https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L63FcP-26NzlRvlR8Ch:bxlz6qh>*
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/22737#issuecomment-368127766>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBcfBc4XvIIH0tYk1onr9A_85BJEOUks5tXx2XgaJpZM4SHJwX>
> .
>

abhimadan commented:
Calling `Recv()` in another goroutine was my original implementation (it had a goroutine leak, but that was fixable). However, I talked to @asubiotto about it and he thought (please correct me if I'm misquoting you Alfonso) that it would be better to have an explicit cancel signal, to be more consistent with the intra-node `ConsumerClosed` message, which has no direct analogue in the FlowStream RPC aside from non-EOF errors, which is somewhat subtle (I didn't name this new API message `ConsumerClosed`, but I can change that to make the connection more clear). I also thought that it would be a bit more flexible to be able to send cancellation signals before the server handler exits (which might be useful for more complex cancellation behaviour in the future), but perhaps that flexibility is unnecessary.

Also, I noticed that the `Recv()` would have some overhead to send the `*ProducerMessage` and `error` to the main goroutine (though I don't know if that overhead is significant). On the other hand, this version doesn't have much channel overhead.

It would require a lot less code to get this working if `Recv` has its own goroutine though. Overall I'm not too committed to either approach, and I don't mind switching back to put `Recv` in its own goroutine if we decide that's a better approach.

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/inbound.go, line 92 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L63H_Qn0k0q21Jq6orA:-L63Zavk8pfohD-ThYJE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa515f0e8070d3a792d8e9b6b546df981837a7d/pkg/sql/distsqlrun/inbound.go#L92)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: either say "If the stream...", or move the comment inside the `if`.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/inbound.go, line 174 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L63HFy_4sGksyXjVfGQ:-L63ZdW_DTcPo5RpZAWa:b393omg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa515f0e8070d3a792d8e9b6b546df981837a7d/pkg/sql/distsqlrun/inbound.go#L174)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can this race with `sendDrainSignalToStreamProducer()`? Is that OK?
</blockquote></details>

You're right, there is a race here, and gRPC doesn't like sending on the same stream from multiple goroutines so it's probably unsafe. I've locked both `sendDrainSignalToStreamProducer` and `sendCancelSignalToStreamProducer` to avoid this. Thanks for noticing this!

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L63Z9ooDkBrOJX4A9Ep:b-alu31e)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
>  it would be better to have an explicit cancel signal, to be more consistent with the intra-node ConsumerClosed message, which has no direct analogue in the FlowStream RPC aside from non-EOF errors

The way I see it, the FlowStream errors are the analogous to `ConsumerClosed`...

> Also, I noticed that the Recv() would have some overhead to send the *ProducerMessage and error to the main goroutine (though I don't know if that overhead is significant)

What I was thinking is that you wouldn't generally send rows to the main goroutine; you'd send them to the `RowReceiver` (`dst`) directly from the goroutine that receives them. The `Recv` goroutine and the RPC handler goroutine would only synchronize for cleanup, not for the hot path of sending rows around. Would that address the overhead?


Of course, this whole thing is unfortunate because gRPC, in collusion with go's networking interfaces, don't give us a way to receive with a cancelable ctx. For reference, what we do in the similar situation for pgwire communication is that we've hacked our own `net.Conn` implementation that transforms reads without timeouts into loops of reads with small timeouts and it checks a context in the loop. Could we maybe do the same for these gRPC connections - is there perhaps a way to override the `net.Conn` that gRPC opens? I think it might be - on the outbound side I believe we already overwrite the `net.Dialer` that gRPC uses to open connections to other nodes. I don't know how stuff works on the inbound side... But maybe there's a hack to be hacked. Look for `readTimeoutConn` and `rpc/context` for setting the dialer on the outbound side.

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L63h-TgCk0xUegXh2Vw:b-lhhdrl)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
The `net.Conn` that a gRPC request uses is given to it by the `net.Listener` we pass to the `grpc.Server.Serve()` method. We already pass in a listener from `CMux.Match()`, but we currently pass `cmux.Any()` when we get the listener that gRPC uses. Unfortunately, it doesn't look like the cmux listeners have the capability to return special `net.Conn`'s; they're just a multiplexing wrapper around a provided `net.Listener`. This would mean that we would have to embed that behaviour in the listener we provide to cmux, but then that listener would have to understand gRPC service and method names and do its own multiplexing so that we can explicitly provide a context to a `readTimeoutConn` (or something similar) depending on the service and method. I suppose making a more complicated cmux would be a possible solution, but as you said, this is probably something gRPC should support rather than being something we have to hack in. Also, it is possible to get a working solution by erroring out the RPC on the server side on cancellation (or doing the more complicated cancellation request handshake I currently have here), so I don't think it's worth investing in making a gRPC-aware multiplexing listener for this purpose, at least right now. (I also haven't really thought about plumbing a context all the way to the server package so that the multiplexer would be able to give it to the `readTimeoutConn`; that would probably be another source of complexity.)

I see your point re: FlowStream errors == ConsumerClosed signals. Since that implementation was simpler and the semantics _are_ well-defined (though a bit unclear IMO, but that can be fixed with a comment or two), I'll go back to doing that, and I'll use your suggestion to reduce channel overhead.

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L6Hn2b6-cORIcXlgeEx:b-qy4l9t)*
<!-- Sent from Reviewable.io -->

Commit:
distsql: Check for cancellation while waiting for inbound streams

Previously, the flow context was checked for cancellation just before
the server sent an RPC to producers. However, if the context was
canceled while blocked on the RPC, the cancellation wouldn't be picked
up until the next producer RPC was sent.

To avoid this problem, the RPC stream is read in its own goroutine, and
the server waits for either context cancellation or the RPC to return,
whichever comes first.

Part of #22650.

Release note: None

abhimadan commented:
Revamped so that the `Recv()` is in a separate goroutine, and it sends rows directly to `dst`. PTAL!

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L6IVxgRBl-9dd6AeahF:bt939np)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/inbound.go, line 45 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6IZK064Ky0ZkQcSp-a:-L6IZK07-HhQ-O2V_c__:b-89hdkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dc3597dc5de9ec0404604567a78c33374141af9/pkg/sql/distsqlrun/inbound.go#L45)):*
> ```Go
> 		log.VEventf(ctx, 1, "inbound stream error: %s", err)
> 		dst.Push(nil, &ProducerMetadata{Err: err})
> 		dst.ProducerDone()
> ```

Ugh, now `ProducerDone()` and `Push()` are called in separate goroutines and they might race. I'll need to adjust this logic so that only one of these goroutines interacts with `dst`.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/inbound.go, line 45 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6IZK064Ky0ZkQcSp-a:-L6IfuFw0rvpOHVn2SZB:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dc3597dc5de9ec0404604567a78c33374141af9/pkg/sql/distsqlrun/inbound.go#L45)):*
<details><summary><i>Previously, abhimadan (Abhishek Madan) wrote…</i></summary><blockquote>

Ugh, now `ProducerDone()` and `Push()` are called in separate goroutines and they might race. I'll need to adjust this logic so that only one of these goroutines interacts with `dst`.
</blockquote></details>

Fixed.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
nit: your release note is unusual. This doesn't fall under (any of the many things) we generally talk about when we talk about performance, and also "cancelations" and "DistSQL flows" don't make sense to most of our readers. I wouldn't add any release node - which doesn't make it any less of a great patch :P

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/inbound.go, line 74 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6ImnK7DRP1Qc1WrrfX:-L6ImnK7DRP1Qc1WrrfY:b9a0lr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L74)):*
> ```Go
> 	// the error to the producer. If useFinalErr is set, the caller must return
> 	// finalErr, even if it's nil.
> 	processProducerMessage := func(msg *ProducerMessage) (err error, useFinalErr bool) {
> ```

please make this a standalone method.

Also, even if you assign names to the return values, please don't use naked returns in the body (which I think you do implicitly when the you let control get to the end of the loop). When we use named ret vals, we only use those names in defered functions.

Also, if you to use named return values (which is not usual), please don't name one of them `err` because of the risk of inadvertently using that name in the method body. For the error just name it `_`. 
`useFinalErr` strikes me as a weird name, which leads to a comment that makes little sense. Call it `consumerClosed`, and explain what it means (if at all), not what the caller should do with it.
Better yet, I don't think this method deserves named ret vals. Just comment something like `The returned bool will be true if the consumer is in state ConsumerClosed`.

---

*[pkg/sql/distsqlrun/inbound.go, line 145 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6IsItY8HfHsVwJmyD7:-L6IsItY8HfHsVwJmyD8:b-4bnzo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L145)):*
> ```Go
> 	// and the RPC goroutine will deadlock. To prevent this, make errChan a
> 	// buffered channel.
> 	errChan := make(chan error, 1)
> ```

This comment left me scratching my head for a bit. Took me a while to understand what it's trying to tell me (some of the wording is vague like "the goroutine sends an error" - to whom?), and after I've figured out what it wants, my first instinct was to say that there's got to be a better way to scope the goroutines that would avoid this buffer and avoid messing with the flow's waitgroup. Only after thinking more I realized that there isn't because... well because the whole problem we're solving is very unfortunate.

How about:
```
There's two goroutines involved in handling the RPC - the current one (the "parent"), which is watching for context cancellation, and a "reader" one that receives messages from the stream. This is all because a stream.Recv() call doesn't react to context cancellation. The idea is that, if the parent detects a canceled context, it will return from this RPC handler, which will cause the stream to be closed.
Because the parent cannot wait for the reader to finish (that being the whole point of the different goroutines), the reader sending an error to the parent might race with the parent finishing. In that case, nobody cares about the reader anymore and so its result channel is buffered.
```

I might find some channel to ask the gRPC-go people if Recv()ing on a stream while returning from the handler is kosher as far as they're concerned.

---

*[pkg/sql/distsqlrun/inbound.go, line 154 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6Irm8d52wCldrkiOjC:-L6Irm8d52wCldrkiOjD:b5uhd6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L154)):*
> ```Go
> 			if err != nil {
> 				if err != io.EOF {
> 					if f.Ctx.Err() != nil {
> ```

I'd remove this special handling of cancellation here; I don't think it serves any purpose.
I've left it alone in #22908 but since you're touching this code, might as well get rid of it.

---

*[pkg/sql/distsqlrun/inbound.go, line 185 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6ImIWW0xN9cuNC5vck:-L6Im2XI6PHWlDPpUR8h:bimszop) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L185)):*
> ```Go
> 	}()
> 
> 	// Check for context cancellation while recv()ing the FlowStream on another
> ```

either stylize as `Recv()ing`, which is the method's name, or better just say "while reading *from* the stream"

---

*[pkg/sql/distsqlrun/inbound.go, line 189 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6IqKf_0buB1NRuWPjH:-L6IqKf_0buB1NRuWPjI:bhpjwjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L189)):*
> ```Go
> 	select {
> 	case <-f.Ctx.Done():
> 		// The flow's context is canceled. Error out the RPC to make the outbox on
> ```

s/Error out the RPC to make the outbox on the producer node cancel its own flow context./Propagate the cancellation to the producer.

In my opinion no comment is necessary here.

---

*[pkg/sql/distsqlrun/inbound.go, line 193 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6IqaPa4UXNRh8GvBPh:-L6IqaPa4UXNRh8GvBPi:bccihi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L193)):*
> ```Go
> 		return sqlbase.NewQueryCanceledError()
> 	case err := <-errChan:
> 		// The producer is done, or the RPC errored out. Return the error to close
> ```

ditto: I would strike the comment.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L6IkOml7D-mlYWsufOZ:bok0s6r)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/inbound.go, line 74 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6ImnK7DRP1Qc1WrrfX:-L6MpxFk3SpZ4mQhIKRN:b-w1ifq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L74)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please make this a standalone method.

Also, even if you assign names to the return values, please don't use naked returns in the body (which I think you do implicitly when the you let control get to the end of the loop). When we use named ret vals, we only use those names in defered functions.

Also, if you to use named return values (which is not usual), please don't name one of them `err` because of the risk of inadvertently using that name in the method body. For the error just name it `_`. 
`useFinalErr` strikes me as a weird name, which leads to a comment that makes little sense. Call it `consumerClosed`, and explain what it means (if at all), not what the caller should do with it.
Better yet, I don't think this method deserves named ret vals. Just comment something like `The returned bool will be true if the consumer is in state ConsumerClosed`.
</blockquote></details>

Made a standalone method, and removed the return value names.

---

*[pkg/sql/distsqlrun/inbound.go, line 145 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6IsItY8HfHsVwJmyD7:-L6Mq1wi8KR4uVuxhqTd:b-cqwh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L145)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

This comment left me scratching my head for a bit. Took me a while to understand what it's trying to tell me (some of the wording is vague like "the goroutine sends an error" - to whom?), and after I've figured out what it wants, my first instinct was to say that there's got to be a better way to scope the goroutines that would avoid this buffer and avoid messing with the flow's waitgroup. Only after thinking more I realized that there isn't because... well because the whole problem we're solving is very unfortunate.

How about:
```
There's two goroutines involved in handling the RPC - the current one (the "parent"), which is watching for context cancellation, and a "reader" one that receives messages from the stream. This is all because a stream.Recv() call doesn't react to context cancellation. The idea is that, if the parent detects a canceled context, it will return from this RPC handler, which will cause the stream to be closed.
Because the parent cannot wait for the reader to finish (that being the whole point of the different goroutines), the reader sending an error to the parent might race with the parent finishing. In that case, nobody cares about the reader anymore and so its result channel is buffered.
```

I might find some channel to ask the gRPC-go people if Recv()ing on a stream while returning from the handler is kosher as far as they're concerned.
</blockquote></details>

Thanks for the great comment! Added.

FWIW, when I run this and cancel a multi-node distsql query, I always get a `rpc error: code = Canceled desc = context canceled` from the inbound `Recv()` after the handler returns.

---

*[pkg/sql/distsqlrun/inbound.go, line 154 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6Irm8d52wCldrkiOjC:-L6Mq55BD9UK6cRFz-S3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L154)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd remove this special handling of cancellation here; I don't think it serves any purpose.
I've left it alone in #22908 but since you're touching this code, might as well get rid of it.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/inbound.go, line 185 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6ImIWW0xN9cuNC5vck:-L6Mq8b09jo6PBSsNGaN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L185)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

either stylize as `Recv()ing`, which is the method's name, or better just say "while reading *from* the stream"
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/inbound.go, line 189 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6IqKf_0buB1NRuWPjH:-L6Mq9Ga5R9l5lV-QIsH:b11x3hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L189)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/Error out the RPC to make the outbox on the producer node cancel its own flow context./Propagate the cancellation to the producer.

In my opinion no comment is necessary here.

</blockquote></details>

Removed the comment.

---

*[pkg/sql/distsqlrun/inbound.go, line 193 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-L6IqaPa4UXNRh8GvBPh:-L6MqBDC6Z2Y2rWqPncw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc7bf4d386d9764a38ce044c6b6c0be0222f7b26/pkg/sql/distsqlrun/inbound.go#L193)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ditto: I would strike the comment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
There's an open gRPC issue that seems to be about the same issue we have here, so we might want to keep tabs on it: https://github.com/grpc/grpc-go/issues/847

---

Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22737#-:-L6N-EcO8Sf-ECegxFKH:b-zdw8u)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
Thanks for working on this!

---

Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/inbound.go, line 146 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22737#-L6NTPDySVNn62DDiNwI:-L6NTPDySVNn62DDiNwJ:b-4taldq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b442bdf1d77487144e9f18ffc38980e8f6d42581/pkg/sql/distsqlrun/inbound.go#L146)):*
> ```Go
> }
> 
> // processProducerMessage is a helper function to read data from the producer
> ```

this function doesn't do any reading.

---

*[pkg/sql/distsqlrun/inbound.go, line 209 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22737#-L6NUd3OcJnygZgXwZRN:-L6NUd3OcJnygZgXwZRO:bj6k1lh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b442bdf1d77487144e9f18ffc38980e8f6d42581/pkg/sql/distsqlrun/inbound.go#L209)):*
> ```Go
> 					// failure to send this signal. This suggests that maybe we want to
> 					// swallow this error...
> 					rs.finalErr = err
> ```

at this point, I think it's time to get rid of this `rs.finalErr`. I think it was always a bad idea to keep track of it, and particularly now it's complicating the code significantly.
I think we should just return the error here to the caller. What do you think?

---

*[pkg/sql/distsqlrun/inbound.go, line 227 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22737#-L6NT7_aH4-tBfYgk_Nt:-L6NT7_aH4-tBfYgk_Nu:b-q1a6l1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b442bdf1d77487144e9f18ffc38980e8f6d42581/pkg/sql/distsqlrun/inbound.go#L227)):*
> ```Go
> }
> 
> type processMessageResult struct {
> ```

nit: I think it'd be more usual to have two return values instead of this type

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22737#-:-L6NUycU7RS4-LLNzUD4:b9c0oiu)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
Thanks for the review!

---

Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/inbound.go, line 146 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22737#-L6NTPDySVNn62DDiNwI:-L6NbqV_VEo5nONVs8Kq:bsn31v5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b442bdf1d77487144e9f18ffc38980e8f6d42581/pkg/sql/distsqlrun/inbound.go#L146)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this function doesn't do any reading.
</blockquote></details>

Right, good point. Did s/read/process.

---

*[pkg/sql/distsqlrun/inbound.go, line 209 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22737#-L6NUd3OcJnygZgXwZRN:-L6Nc20Y2oFzNEtxnufA:bx0g632) ([raw file](https://github.com/cockroachdb/cockroach/blob/b442bdf1d77487144e9f18ffc38980e8f6d42581/pkg/sql/distsqlrun/inbound.go#L209)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

at this point, I think it's time to get rid of this `rs.finalErr`. I think it was always a bad idea to keep track of it, and particularly now it's complicating the code significantly.
I think we should just return the error here to the caller. What do you think?
</blockquote></details>

I agree, and the rationale in your comment about it in the code makes sense. It's probably better to just log it.

---

*[pkg/sql/distsqlrun/inbound.go, line 227 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22737#-L6NT7_aH4-tBfYgk_Nt:-L6NcDluPkkfgpAtg6vJ:b-bfjw30) ([raw file](https://github.com/cockroachdb/cockroach/blob/b442bdf1d77487144e9f18ffc38980e8f6d42581/pkg/sql/distsqlrun/inbound.go#L227)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I think it'd be more usual to have two return values instead of this type
</blockquote></details>

I agree, but I made this struct to appease the linter. It didn't like the error being the first return value, but in this case that made sense because the bool was supplementary information and the error was the main value. The struct feels a bit excessive, but I wasn't sure how else to deal with that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22737#-:-L6NbjssO4V_UKFDoa7b:b-9qdoz3)*
<!-- Sent from Reviewable.io -->
